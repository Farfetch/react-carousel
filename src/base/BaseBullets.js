import Bullet from '#components/Bullet';
import BulletsInfiniteWrapper from '#components/BulletsInfiniteWrapper';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

class BaseBullets extends PureComponent {
    static getDerivedStateFromProps(props, state) {
        const bullets = Math.ceil(props.itemsLength / props.itemsToScroll);

        const activeBullet = Math.ceil(props.activeItem / props.itemsToScroll);

        return {
            isInfinite: bullets > state.maxBullets,
            translationValue: props.isRTL ? 1.8 : -1.8,
            itemsToScroll: props.itemsToScroll,
            bullets,
            activeBullet,
        };
    }

    constructor(props) {
        super(props);

        this.state = {
            maxBullets: 5,
            translationValue: 0,
            isInfinite: false,
            itemsToScroll: 1,
            bullets: 5,
            activeBullet: 1,
        };
    }

    render() {
        const {
            activeItem,
            itemsLength,
            isRTL,
            itemsToScroll,
            ...otherProps
        } = this.props;
        const { isInfinite } = this.state;

        return (
            <div
                aria-hidden="true"
                data-infinite={isInfinite}
                data-active={activeItem}
                {...otherProps}
            >
                {isInfinite ? (
                    <BulletsInfiniteWrapper
                        style={{
                            transform: `translate(${this.getTranslationValue()}rem)`,
                        }}
                    >
                        {this.getBullets()}
                    </BulletsInfiniteWrapper>
                ) : (
                    this.getBullets()
                )}
            </div>
        );
    }

    getSecondaryBullets() {
        const { activeBullet, bullets } = this.state;

        return {
            // Secondary bullet is the previous one, unless active bullet is the first
            first: activeBullet === 0 ? 2 : activeBullet - 1,
            // Other secondary bullet is the next one, unless active bullet is the last one
            second:
                activeBullet === bullets ? activeBullet - 2 : activeBullet + 1,
        };
    }

    getTranslationValue() {
        const {
            translationValue,
            maxBullets,
            bullets,
            activeBullet,
        } = this.state;

        // If active slide is in first, second or third position translation doesn't happen
        if (bullets <= maxBullets || activeBullet <= 2) {
            return 0;
        }
        // If active slide is one of the last three positions translation stays the same
        if (activeBullet > bullets - 3) {
            return (bullets - maxBullets) * translationValue;
        }

        // Generic case where the bullets are translated
        return (activeBullet - 2) * translationValue;
    }

    getBullets() {
        const { isInfinite, activeBullet, bullets } = this.state;
        const secondaryBullets = this.getSecondaryBullets();
        const isSecondary = (index) =>
            secondaryBullets.first === index ||
            secondaryBullets.second === index;

        const itemsNodes = [];

        for (let i = 0; i < bullets; i++) {
            itemsNodes.push(
                <Bullet
                    key={i}
                    data-infinite={isInfinite}
                    data-secondary={isSecondary(i)}
                    data-active={activeBullet === i}
                    data-tstid="bullet"
                />
            );
        }

        return itemsNodes;
    }
}

BaseBullets.propTypes = {
    /* Index of active item - start on 0 */
    activeItem: PropTypes.number.isRequired,
    /* Items length */
    itemsLength: PropTypes.number.isRequired,
    /* Navigation direction  */
    isRTL: PropTypes.bool,
    /* SliderNav theme */
    itemsToScroll: PropTypes.number,
};

BaseBullets.defaultProps = {
    itemsToScroll: 1,
    theme: {},
};

export default BaseBullets;
