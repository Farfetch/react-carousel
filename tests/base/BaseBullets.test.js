import { shallow } from 'enzyme';
import BaseBullets from '../../src/base/BaseBullets';
import React from 'react';
import serializer from 'enzyme-to-json/serializer';

beforeAll(() => {
    expect.addSnapshotSerializer(serializer);
});

afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
});

const mockPropsDefault = {
    activeItem: 0,
    itemsLength: 3,
};

const mockPropsInfinite = {
    ...mockPropsDefault,
    itemsLength: 10,
};

describe('<BaseBullets> - props for default', () => {
    it('should render correctly the basic', () => {
        const wrapper = shallow(<BaseBullets {...mockPropsDefault} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly with correct state', () => {
        const wrapper = shallow(<BaseBullets {...mockPropsDefault} />);

        const expectedState = {
            maxBullets: 5,
            translationValue: -1.8,
            isInfinite: false,
            itemsToScroll: 1,
            activeBullet: 0,
            bullets: 3,
        };

        expect(wrapper.state()).toEqual(expectedState);
    });

    it('should render correctly with correct state after update props', () => {
        const wrapper = shallow(<BaseBullets {...mockPropsDefault} />, {
            disableLifecycleMethods: false,
        });

        wrapper.setProps({ isRTL: true, itemsLength: 4 });

        const expectedState = {
            maxBullets: 5,
            translationValue: 1.8,
            isInfinite: false,
            itemsToScroll: 1,
            activeBullet: 0,
            bullets: 4,
        };

        expect(wrapper.state()).toEqual(expectedState);
    });
});

describe('<BaseBullets> - props for infinite bullets', () => {
    it('should render correctly when activeItem is the first one', () => {
        const wrapper = shallow(<BaseBullets {...mockPropsInfinite} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly when activeItem is the second one', () => {
        const wrapper = shallow(
            <BaseBullets {...mockPropsInfinite} activeItem={1} />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly when activeItem is the last one', () => {
        const wrapper = shallow(
            <BaseBullets
                {...mockPropsInfinite}
                activeItem={mockPropsInfinite.itemsLength - 1}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly when activeItem is the penultimate one', () => {
        const wrapper = shallow(
            <BaseBullets
                {...mockPropsInfinite}
                activeItem={mockPropsInfinite.itemsLength - 2}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });
});

describe('<BaseBullets> - lifecycling for infinite bullets', () => {
    it('should update secondaryBullets and bullets when newProp.activeItem is not equal to the current one', () => {
        const spyGetSecondaryBullets = jest.spyOn(
            BaseBullets.prototype,
            'getSecondaryBullets'
        );
        const wrapper = shallow(<BaseBullets {...mockPropsInfinite} />, {
            disableLifecycleMethods: false,
        });

        const expectedBefore = {
            first: 2,
            second: 1,
        };

        expect(spyGetSecondaryBullets).toHaveBeenCalledTimes(1);
        expect(spyGetSecondaryBullets).toHaveReturnedWith(expectedBefore);

        const mockIndex = 1;
        const expectedAfter = {
            first: 0,
            second: 2,
        };

        wrapper.setProps({
            activeItem: mockIndex,
        });

        expect(wrapper.prop('data-active')).toBe(mockIndex);
        expect(spyGetSecondaryBullets).toHaveBeenCalledTimes(2);
        expect(spyGetSecondaryBullets).toHaveReturnedWith(expectedAfter);
        expect(wrapper).toMatchSnapshot();
    });

    it('should not update secondaryBullets and bullets when newProp.activeItem is equal to the current one', () => {
        const spyGetSecondaryBullets = jest.spyOn(
            BaseBullets.prototype,
            'getSecondaryBullets'
        );

        const wrapper = shallow(<BaseBullets {...mockPropsInfinite} />, {
            disableLifecycleMethods: false,
        });

        wrapper.setProps({
            activeItem: mockPropsInfinite.activeItem,
        });

        expect(wrapper.prop('data-active')).toBe(mockPropsInfinite.activeItem);
        expect(spyGetSecondaryBullets).toHaveBeenCalledTimes(1);
    });

    it('should update isInfinite when newProps.itemsLength is changed', () => {
        const wrapper = shallow(<BaseBullets {...mockPropsDefault} />);

        wrapper.setProps({
            itemsLength: 4,
        });

        const expectedState = {
            maxBullets: 5,
            translationValue: -1.8,
            isInfinite: false,
            itemsToScroll: 1,
            activeBullet: 0,
            bullets: 4,
        };

        expect(wrapper.state()).toEqual(expectedState);
    });
});
