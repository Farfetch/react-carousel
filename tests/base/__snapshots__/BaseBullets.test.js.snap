// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseBullets> - lifecycling for infinite bullets should update secondaryBullets and bullets when newProp.activeItem is not equal to the current one 1`] = `
<div
  aria-hidden="true"
  data-active={1}
  data-infinite={true}
>
  <BulletsInfiniteWrapper
    style={
      Object {
        "transform": "translate(0rem)",
      }
    }
  >
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="0"
    />
    <Bullet
      data-active={true}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="1"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="2"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="3"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="4"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="5"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="6"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="7"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="8"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="9"
    />
  </BulletsInfiniteWrapper>
</div>
`;

exports[`<BaseBullets> - props for default should render correctly the basic 1`] = `
<div
  aria-hidden="true"
  data-active={0}
  data-infinite={false}
>
  <Bullet
    data-active={true}
    data-infinite={false}
    data-secondary={false}
    data-tstid="bullet"
    key="0"
  />
  <Bullet
    data-active={false}
    data-infinite={false}
    data-secondary={true}
    data-tstid="bullet"
    key="1"
  />
  <Bullet
    data-active={false}
    data-infinite={false}
    data-secondary={true}
    data-tstid="bullet"
    key="2"
  />
</div>
`;

exports[`<BaseBullets> - props for infinite bullets should render correctly when activeItem is the first one 1`] = `
<div
  aria-hidden="true"
  data-active={0}
  data-infinite={true}
>
  <BulletsInfiniteWrapper
    style={
      Object {
        "transform": "translate(0rem)",
      }
    }
  >
    <Bullet
      data-active={true}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="0"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="1"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="2"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="3"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="4"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="5"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="6"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="7"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="8"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="9"
    />
  </BulletsInfiniteWrapper>
</div>
`;

exports[`<BaseBullets> - props for infinite bullets should render correctly when activeItem is the last one 1`] = `
<div
  aria-hidden="true"
  data-active={9}
  data-infinite={true}
>
  <BulletsInfiniteWrapper
    style={
      Object {
        "transform": "translate(-9rem)",
      }
    }
  >
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="0"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="1"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="2"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="3"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="4"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="5"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="6"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="7"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="8"
    />
    <Bullet
      data-active={true}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="9"
    />
  </BulletsInfiniteWrapper>
</div>
`;

exports[`<BaseBullets> - props for infinite bullets should render correctly when activeItem is the penultimate one 1`] = `
<div
  aria-hidden="true"
  data-active={8}
  data-infinite={true}
>
  <BulletsInfiniteWrapper
    style={
      Object {
        "transform": "translate(-9rem)",
      }
    }
  >
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="0"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="1"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="2"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="3"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="4"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="5"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="6"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="7"
    />
    <Bullet
      data-active={true}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="8"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="9"
    />
  </BulletsInfiniteWrapper>
</div>
`;

exports[`<BaseBullets> - props for infinite bullets should render correctly when activeItem is the second one 1`] = `
<div
  aria-hidden="true"
  data-active={1}
  data-infinite={true}
>
  <BulletsInfiniteWrapper
    style={
      Object {
        "transform": "translate(0rem)",
      }
    }
  >
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="0"
    />
    <Bullet
      data-active={true}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="1"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={true}
      data-tstid="bullet"
      key="2"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="3"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="4"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="5"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="6"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="7"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="8"
    />
    <Bullet
      data-active={false}
      data-infinite={true}
      data-secondary={false}
      data-tstid="bullet"
      key="9"
    />
  </BulletsInfiniteWrapper>
</div>
`;
